### ISSUE TITLE
RuntimeError: Cannot find new config in BaseSearcher, even after 100 trials

### ISSUE THREAD

**Issue by talhaanwarch** [awslabs/autogluon#1141]
I am getting `RuntimeError: Cannot find new config in BaseSearcher, even after 100 trials` when using `medium_quality_faster_inference` as preset in `ImagePredictor`.

Colab notebook 
https://colab.research.google.com/drive/1ekpD-r29DqtDiPXCy9zHHA8LTKLeTDrN?usp=sharing

Error log
```
<ipython-input-5-b3c4aba9b9da> in <module>()
      1 predictor = ImagePredictor(verbosity=1)
      2 # since the original dataset does not provide validation split, the `fit` function splits it randomly with 90/10 ratio
----> 3 predictor.fit(train_dataset,tuning_data=test_dataset, presets=['medium_quality_faster_inference'],hyperparameters={'epochs': 1})  # you can trust the default config, we reduce the # epoch to save some build time

1 frames
/usr/local/lib/python3.7/dist-packages/autogluon/vision/configs/presets_configs.py in _call(*args, **kwargs)
     13         def _call(*args, **kwargs):
     14             gargs, gkwargs = set_presets(preset_name, *args, **kwargs)
---> 15             return f(*gargs, **gkwargs)
     16         return _call
     17     return _unpack_inner

/usr/local/lib/python3.7/dist-packages/autogluon/vision/predictor/predictor.py in fit(self, train_data, tuning_data, time_limit, presets, hyperparameters, **kwargs)
    375                 self._classifier = task.fit(train_data, tuning_data, 1 - holdout_frac, random_state)
    376             if err.exc_value is not None:
--> 377                 raise RuntimeError(err.exc_value + err.hint)
    378         self._classifier._logger.setLevel(log_level)
    379         self._classifier._logger.propagate = True

RuntimeError: Cannot find new config in BaseSearcher, even after 100 trials
```

**Comment by talhaanwarch** [comment]
Please check this notebook, 
https://colab.research.google.com/drive/1lo9FIkiubR28ZOxTAkT7qMcB4-ECU5iU?usp=sharing
The error is quite weird.
By default preset is `['medium_quality_faster_train']`, if I define it explicitly, I got the error. Otherwise, I did not

**Comment by Innixma** [comment]
Thanks for checking this and providing colab notebooks!

I believe this is because there is no actual search space specified in `medium_quality_faster_train`. This is the config:

```
    medium_quality_faster_train={
        'hyperparameters': {
            'model': 'resnet50_v1b',
            'lr': 0.01,
            'batch_size': 64,
            'epochs': 50,
            'early_stop_patience': 5
            },
        'hyperparameter_tune_kwargs': {
            'num_trials': 8,
            'search_strategy': 'random',
        },
        'time_limit': 1*3600,
    },
```

As a comparison, here is the `best_quality` config: 

```
    best_quality={
        'hyperparameters': {
            'model': Categorical('resnet50_v1b', 'resnet101_v1d', 'resnest200'),
            'lr': Real(1e-5, 1e-2, log=True),
            'batch_size': Categorical(8, 16, 32, 64, 128),
            'epochs': 200,
            'early_stop_patience': 50
            },
        'hyperparameter_tune_kwargs': {
            'num_trials': 1024,
            'search_strategy': 'bayesopt',
        },
        'time_limit': 12*3600,
    },
```

As can be seen, `best_quality` searches over model, lr, and batch_size. You would get a similar error by specifying `best_quality` and specifying fixed values for model, lr, and batch_size explicitly (overriding the config values).

> The error is quite weird.
By default preset is ['medium_quality_faster_train'], if I define it explicitly, I got the error. Otherwise, I did not

The documentation is not correct in this case, while it claims it is the default, it is not. This is likely because it was copy pasted from Tabular's preset documentation, whose config for medium_quality_faster_train is:

```
medium_quality_faster_train={'auto_stack': False},
```

Which is already the default (`auto_stack=False` by default).

@zhreshold Could we update the documentation and fix the config? Perhaps remove the `hyperparameter_tune_kwargs` dict from the preset so it doesn't tune, or add a search space to it.

**Comment by talhaanwarch** [comment]
So if i have to use define presets explicitly, I should not define hyper-parameters explicitly?

**Comment by Innixma** [comment]
You can still do so, I'm just referring to how the logic works (if you specify a given hyperparameter value, that value will override what is set in the preset). The bug is that if there are no hyperparameters with searchspaces and hyperparmeter tuning is enabled, the code will crash with the cryptic error you mentioned. Two things need to be fixed: 1. Make the error less cryptic. 2. Make `medium_quality_faster_train` a valid configuration.

**Comment by talhaanwarch** [comment]
But if you check, `medium_quality_faster_inference`, according to my understanding, there are hyper-parameters in search space, so why it raises the above issue?
```
‘model’: Categorical(‘resnet18_v1b’, ‘mobilenetv3_small’), 
‘lr’: Categorical(0.01, 0.005, 0.001),
 ‘batch_size’: Categorical(64, 128),
 ‘epochs’: Categorical(50, 100), 
‘early_stop_patience’: 10 },
```

**Comment by Innixma** [comment]
Oh, my mistake. I misread your initial comment as `medium_quality_faster_train` instead of `medium_quality_faster_inference`.

I'm unsure why `medium_quality_faster_inference` would crash. @zhreshold It's possible that the hyperparameter overrides are not working correctly for Vision and they are overwriting the entire `hyperparameters` dictionary instead of individual keys within the dictionary.

Thanks again for pointing this out. Apologies for all the bugs! It seems the preset functionality in Vision may have some differences in implementation compared to Tabular's presets. We will plan to fix them for v0.3.0 release.

**Comment by talhaanwarch** [comment]
First, i thought this issue is because of hyper-parameters overriding, but it is not. check this notebook
https://colab.research.google.com/drive/1SshDbvdCSzvRpxDc5qRsvxM03NfKEUqp?usp=sharing

**Comment by talhaanwarch** [comment]
Btw thanks @Innixma you are doing great. AG is quite an amazing package.

**Comment by Innixma** [comment]
Further diving into this problem and its caused by running out of different configurations to search. If all possible hyperparameter spaces in the search space have been checked, this error occurs. Ideally, it should simply stop the search process early and return the best configuration found.

**Comment by zhreshold** [comment]
closed with #1164
